core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -792,7 +801,10 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
       val predicatesMatchingAddedFiles = ExpressionSet(readPredicates).iterator.flatMap { p =>
         val conflictingFile = DeltaLog.filterFileList(
           metadata.partitionSchema,
-          addedFilesToCheckForConflicts.toDF(), p :: Nil).as[AddFile].take(1)
+          addedFilesToCheckForConflicts.filter(isDataChangeOrConflict).toDF(),
+          p :: Nil)
+          .as[AddFile]
```

Comment:
That looks as ugly as before but...you can't do much about that. Sorry for the never-ending begging for change. It's OK.